### ISSUE TITLE
/dev/zvol/dataset and /dev/dataset links are missing for some zvols

### ISSUE THREAD

**Issue by rugubara** [openzfs/zfs#12484]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| Gentoo
Distribution Version	| rolling/~x86_64
Kernel Version	| 5.13.11
Architecture	| x86_64
OpenZFS Version	| 2.1.0

### Describe the problem you're observing
some zvols links don't appear in /dev/zvol, although the /dev/zd### device is present.  
### Describe how to reproduce the problem
I'm not sure when/why this started , but I experience it every boot on my system.
When I issue ``` udevadm settle && udevadm trigger``` a different set of zvols may appear/disappear in /dev/zvol
 
### Include any warning/errors/backtraces from the system logs

```
KERNEL[3918.736693] change   /devices/virtual/block/zd0 (block)
KERNEL[3918.736708] change   /devices/virtual/block/zd0/zd0p1 (block)
KERNEL[3918.736722] change   /devices/virtual/block/zd0/zd0p2 (block)
KERNEL[3918.736736] change   /devices/virtual/block/zd0/zd0p3 (block)
KERNEL[3918.736750] change   /devices/virtual/block/zd0/zd0p4 (block)
KERNEL[3918.736763] change   /devices/virtual/block/zd112 (block)
KERNEL[3918.736775] change   /devices/virtual/block/zd128 (block)
KERNEL[3918.736788] change   /devices/virtual/block/zd144 (block)
KERNEL[3918.736801] change   /devices/virtual/block/zd16 (block)
KERNEL[3918.736816] change   /devices/virtual/block/zd16/zd16p1 (block)
KERNEL[3918.736838] change   /devices/virtual/block/zd16/zd16p2 (block)
KERNEL[3918.736857] change   /devices/virtual/block/zd160 (block)
KERNEL[3918.736876] change   /devices/virtual/block/zd176 (block)
KERNEL[3918.736895] change   /devices/virtual/block/zd192 (block)
KERNEL[3918.736913] change   /devices/virtual/block/zd208 (block)
KERNEL[3918.737007] change   /devices/virtual/block/zd224 (block)
KERNEL[3918.737024] change   /devices/virtual/block/zd240 (block)
KERNEL[3918.737038] change   /devices/virtual/block/zd32 (block)
KERNEL[3918.737052] change   /devices/virtual/block/zd48 (block)
KERNEL[3918.737069] change   /devices/virtual/block/zd64 (block)
KERNEL[3918.737086] change   /devices/virtual/block/zd80 (block)
KERNEL[3918.737103] change   /devices/virtual/block/zd96 (block)
UDEV  [3918.842945] change   /devices/virtual/block/zd240 (block)
UDEV  [3918.846885] change   /devices/virtual/block/zd176 (block)
UDEV  [3918.849689] change   /devices/virtual/block/zd16 (block)
UDEV  [3918.849729] change   /devices/virtual/block/zd64 (block)
UDEV  [3918.852253] change   /devices/virtual/block/zd128 (block)
UDEV  [3918.853294] change   /devices/virtual/block/zd48 (block)
UDEV  [3918.854459] change   /devices/virtual/block/zd80 (block)
UDEV  [3918.857068] change   /devices/virtual/block/zd224 (block)
UDEV  [3918.857497] change   /devices/virtual/block/zd144 (block)
UDEV  [3918.857657] change   /devices/virtual/block/zd96 (block)
UDEV  [3918.857752] change   /devices/virtual/block/zd16/zd16p1 (block)
UDEV  [3918.857799] change   /devices/virtual/block/zd160 (block)
UDEV  [3918.858854] change   /devices/virtual/block/zd0 (block)
UDEV  [3918.861916] change   /devices/virtual/block/zd192 (block)
UDEV  [3918.864396] change   /devices/virtual/block/zd16/zd16p2 (block)
UDEV  [3918.868286] change   /devices/virtual/block/zd112 (block)
UDEV  [3918.871430] change   /devices/virtual/block/zd0/zd0p2 (block)
UDEV  [3918.871500] change   /devices/virtual/block/zd0/zd0p4 (block)
UDEV  [3918.871746] change   /devices/virtual/block/zd0/zd0p3 (block)
UDEV  [3918.884162] change   /devices/virtual/block/zd0/zd0p1 (block)
UDEV  [3918.886678] change   /devices/virtual/block/zd208 (block)
UDEV  [3918.894483] change   /devices/virtual/block/zd32 (block)



PF16W6Y2 ~ # zfs list -r fast/VM/storage/gentoo-lab
NAME                            USED  AVAIL     REFER  MOUNTPOINT
fast/VM/storage/gentoo-lab      120G   232G       96K  legacy
fast/VM/storage/gentoo-lab/D1  20.1G   232G     20.1G  -
fast/VM/storage/gentoo-lab/D2  20.1G   232G     20.1G  -
fast/VM/storage/gentoo-lab/D3  20.1G   232G     20.1G  -
fast/VM/storage/gentoo-lab/D4  20.1G   232G     20.1G  -
fast/VM/storage/gentoo-lab/D5  20.1G   232G     20.1G  -
fast/VM/storage/gentoo-lab/D6  20.1G   232G     20.1G  -

PF16W6Y2 ~ # ls -l /dev/zvol/fast/VM/storage/gentoo-lab/
total 0
lrwxrwxrwx 1 root root 20 авг 17 15:22 D1 -> ../../../../../zd160
lrwxrwxrwx 1 root root 20 авг 17 15:22 D2 -> ../../../../../zd192
lrwxrwxrwx 1 root root 20 авг 17 15:22 D4 -> ../../../../../zd224

PF16W6Y2 ~ # systemctl --version
systemd 249 (249)
+PAM -AUDIT -SELINUX -APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID -CURL -ELFUTILS -FIDO2 +IDN2 -IDN -IPTC +KMOD -LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY -P11KIT -QRENCODE -BZIP2 +LZ4 -XZ -ZLIB +ZSTD -XKBCOMMON +UTMP +SYSVINIT default-hierarchy=hybrid

```

**Comment by rugubara** [comment]
Today I managed to create all the /dev/zvol links with the following command:
```find /sys/devices/virtual/block -name 'zd*' -print |grep -v p|xargs -n1 udevadm test```

I'll try to issue ```udevadm test``` in parallel to simulate the behaviour of udev

**Comment by dioni21** [comment]
Just to be sure, you do NOT have ASAN or any other debug enabled in build, right?
Original RPMs?

**Comment by rugubara** [comment]
Gentoo builds all packages from sources before installation. AFAIK, there is no debug enabled. How can I verify?

**Comment by hiroshimatsuo** [comment]
My Fedora 34 (and 33) machine,
kernel:  5.13.12
openzfs:  2.1.0
has this problem too.

My adhoc workaround is export/import zpool.
I don't know why there is the difference between udevs during booting and after.

```
[root@tiger ~]# ls -l /dev/zvol/
total 0
lrwxrwxrwx 1 root root 8 Aug 23 04:46 Unable -> ../zd0p2
[root@tiger ~]# ls -l /dev/zd*
brw-rw---- 1 root disk 230, 0 Aug 23 04:46 /dev/zd0
brw-rw---- 1 root disk 230, 1 Aug 23 04:46 /dev/zd0p1
brw-rw---- 1 root disk 230, 2 Aug 23 04:46 /dev/zd0p2
[root@tiger ~]# zfs list
NAME                   USED  AVAIL     REFER  MOUNTPOINT
mypool                3.84T  3.30T      128K  /mypool
mypool/data           1.11T  3.30T     1.11T  /mypool/data
mypool/export          128K  3.30T      128K  /mypool/export
mypool/vol0           2.73T  5.45T      589G  -
pool500g              29.4G   870G     25.5K  /pool500g
pool500g/timemachine  29.4G   870G     29.4G  /pool500g/timemachine
[root@tiger ~]#
[root@tiger ~]#
[root@tiger ~]# zpool export mypool
[root@tiger ~]# zpool import mypool
[root@tiger ~]# ls -l /dev/zvol/
total 0
drwxr-xr-x 2 root root 100 Aug 23 04:50 mypool
[root@tiger ~]# ls -l /dev/zvol/mypool/
total 0
lrwxrwxrwx 1 root root  9 Aug 23 04:50 vol0 -> ../../zd0
lrwxrwxrwx 1 root root 11 Aug 23 04:50 vol0-part1 -> ../../zd0p1
lrwxrwxrwx 1 root root 11 Aug 23 04:50 vol0-part2 -> ../../zd0p2
[root@tiger ~]#
```

In my memory, kernel 5.11 (with OpenZFS 2.1.0-rc4(?) or 2.0.5(?))
ran properly before upgraded to 5.12.

**Comment by rugubara** [comment]
> My adhoc workaround is export/import zpool.
My pool has /. I can't apply you workaround...

**Comment by jgottula** [comment]
This sounds like it's probably the same as #12301 (a kernel 5.13+ compat problem).

**Comment by rugubara** [comment]
Indeed this is a duplicate of #12301
